File: lecture_notes/09 functions/2-scope.cpp
#include <algorithm>
#include <functional>
#include <iostream>
#include <vector>

using namespace std;

void f1();

template<class T>
void print_vector(const vector<T>& v, const string& comment) {
    cout << comment << " [ ";
    for (const auto& e : v) {
        cout << e << " ";
    }
    cout << "]" << endl;
}

int main() {
    vector<string> cities {"Athens", "Roma", "London", "Beijing", "Kiev", "Riga"};
    auto f = [](const string& s) {
        return s.length();
    };

    f1();

    print_vector(cities, "cities:");
    vector<size_t> modified_cities(cities.size());
    transform(cities.begin(), cities.end(), modified_cities.begin(), f);
    print_vector(modified_cities, "modified cities:");

    return 0;
}

void f1() {
    vector<string> cities {"Athens", "Roma"};
    auto f = [](string s) {
        transform(s.begin(), s.end(), s.begin(), ::toupper);
        return s;
    };

    print_vector(cities, "cities:");

    vector<string> modified_cities(cities.size());
    transform(cities.begin(), cities.end(), modified_cities.begin(), f);
    print_vector(modified_cities, "upper cities:");

    {
        auto f = [](string s) {
            transform(s.begin(), s.end(), s.begin(), ::tolower);
            return s;
        };
        vector<string> modified_cities(cities.size());
        transform(cities.begin(), cities.end(), modified_cities.begin(), f);
        print_vector(cities, "cities:");
        print_vector(modified_cities, "lower cities:");

    }
    {
        vector<string> cities {"London", "Beijing", "Kiev"};
        print_vector(cities, "cities:");
        vector<string> modified_cities(cities.size());
        transform(cities.begin(), cities.end(), modified_cities.begin(), f);
        print_vector(modified_cities, "upper cities:");
    }
}

File: lecture_notes/09 functions/2 - scope.py
cities = ['Athens', 'Roma', 'London', 'Beijing', 'Kiev', 'Riga']


def to_upper_case(list):
    for city in range(len(list)):
        list[city] = list[city].upper()
    return list


print(['Athens', 'Roma'])
print(to_upper_case(['Athens', 'Roma']))


def to_lower_case(list):
    for city in range(len(list)):
        list[city] = list[city].lower()
    return list


print(['Athens', 'Roma'])
print(to_lower_case(['Athens', 'Roma']))

print(['London', 'Beijing', 'Kiev'])
print(to_upper_case(['London', 'Beijing', 'Kiev']))


def count_len(cities):
    return [len(city) for city in cities]


print(cities)
print(count_len(cities))
